demo-app/src/main.rs

```diff
@@ -0,0 +1,253 @@
+use sovereign_sdk::{
+    core::{
+        mocks::MockProof,
+        traits::{BatchTrait, CanonicalHash, TransactionTrait},
+    },
+    jmt,
+    serial::Decode,
+    stf::{ConsensusSetUpdate, OpaqueAddress, StateTransitionFunction},
+};
+
+// we should re-export anyhow from the modules_api
+
+#[derive(Debug, PartialEq, Eq, Clone, BorshDeserialize, BorshSerialize)]
+pub struct DemoTransaction<C: Context> {
+    mock_signature: MockSignature,
+    msg: Vec<u8>,
+    phantom_context: PhantomData<C>,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, BorshDeserialize, BorshSerialize)]
+pub struct MockSignature {
+    pub_key: MockPublicKey,
+    msg_hash: [u8; 32],
+}
+
+pub struct Demo {
+    current_storage: JmtStorage,
+}
+
+impl StateTransitionFunction for Demo {
+    type StateRoot = jmt::RootHash;
+
+    type ChainParams = ();
+
+    type Transaction = DemoTransaction<MockContext>;
+
+    type Batch = Batch<MockContext>;
+
+    type Proof = MockProof;
+
+    type MisbehaviorProof = ();
+
+    fn init_chain(&mut self, _params: Self::ChainParams) {
+        Runtime::<MockContext>::genesis(self.current_storage.clone())
+            .expect("module initialization must succeed");
+        self.current_storage.finalize();
+    }
+
+    fn begin_slot(&self) {}
+
+    fn apply_batch(
+        &self,
+        batch: Self::Batch,
+        sequencer: &[u8],
+        _misbehavior_hint: Option<Self::MisbehaviorProof>,
+    ) -> Result<
+        Vec<Vec<sovereign_sdk::stf::Event>>,
+        sovereign_sdk::stf::ConsensusSetUpdate<OpaqueAddress>,
+    > {
+        let mut storage = self.current_storage.clone();
+        let mut events = Vec::new();
+        for tx in batch.take_transactions() {
+            // Do mock signature verification
+            // We just check that the signature hash matches the tx hash
+            let expected_hash = tx.mock_signature.msg_hash;
+            let found_hash = <MockContext as Spec>::Hasher::hash(&tx.msg);
+            // If the (mock) signature is invalid, the sequencer is malicious. Slash them.
+            if expected_hash != found_hash {
+                return Err(ConsensusSetUpdate::slashing(sequencer));
+            }
+
+            if let Ok(msg) = RuntimeCall::<MockContext>::decode(&mut std::io::Cursor::new(tx.msg)) {
+                let ctx = MockContext::new(tx.mock_signature.pub_key);
+                let tx_result = msg.dispatch_call(storage.clone(), &ctx);
+                match tx_result {
+                    Ok(resp) => {
+                        events.push(resp.events);
+                        storage.merge();
+                    }
+                    Err(_) => {
+                        storage.merge_reads_and_discard_writes();
+                    }
+                }
+            } else {
+                // If the serialization is invalid, the sequencer is malicious. Slash them
+                return Err(ConsensusSetUpdate::slashing(sequencer));
+            }
+        }
```

Comment:
Should we save db here:`storage.finalize()`?

Reply:
I think we want to finalize during `end_slot` for better throughput